sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala

```diff
@@ -161,17 +166,23 @@ private[sql] object ArrowConverters extends Logging {
         val writeChannel = new WriteChannel(Channels.newChannel(out))
 
         var rowCount = 0L
+        var estimatedBatchSize = arrowSchemaSize
         Utils.tryWithSafeFinally {
-          while (rowIter.hasNext && (maxRecordsPerBatch <= 0 || rowCount < maxRecordsPerBatch)) {
+          // Always write the schema.
+          MessageSerializer.serialize(writeChannel, arrowSchema)
+
+          // Always write the first row.
+          while (rowIter.hasNext && (rowCount == 0 || estimatedBatchSize < maxBatchSize)) {
             val row = rowIter.next()
             arrowWriter.write(row)
+            estimatedBatchSize += row.asInstanceOf[UnsafeRow].getSizeInBytes
```

Comment:
refer to how the size is computed in [BroadcastExchange](https://github.com/apache/spark/blob/e17d8ecabcad6e84428752b977120ff355a4007a/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala#L150-L158) 

but not 100% sure, should I use this instead?
```
row match {
   case unsafe: UnsafeRow => estimatedBatchSize += unsafe.getSizeInBytes
   case _ => estimatedBatchSize += SizeEstimator.estimate(row)
}
```

cc @HyukjinKwon

Reply:
The size of message should be based on Arrow but we are only able to know the size of the batch when Arrow batch is created.

So I am fine with the current approach. I do believe that `UnsafeRow` has bigger size than `ArrowBatch` in general.

One nit would be we should probably set the lower size in `maxBatchSize` to be conservative. For example, `maxBatchSize * 0.7`

Reply:
will update `maxBatchSize * 0.7`